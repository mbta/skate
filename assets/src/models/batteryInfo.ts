export interface BatteryInfo {
  chargePercentage: number;
}
