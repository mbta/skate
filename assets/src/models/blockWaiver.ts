import { AlertIconStyle } from "../components/iconAlertCircle"
import { BlockWaiver, VehicleOrGhost } from "../realtime"
import { now } from "../util/dateTime"
import { isGhost, isLateVehicleIndicator } from "./vehicle"

export enum CurrentFuturePastType {
  Current = 1,
  Future,
  Past,
}

export const currentFuturePastType = ({
  startTime,
  endTime,
}: BlockWaiver): CurrentFuturePastType => {
  const nowDate: Date = now()

  if (startTime > nowDate) {
    return CurrentFuturePastType.Future
  } else if (endTime < nowDate) {
    return CurrentFuturePastType.Past
  } else {
    return CurrentFuturePastType.Current
  }
}

export const hasBlockWaiver = ({ blockWaivers }: VehicleOrGhost): boolean =>
  blockWaivers.length !== 0

export const hasCurrentBlockWaiver = ({
  blockWaivers,
}: VehicleOrGhost): boolean =>
  blockWaivers.some(
    blockWaiver =>
      currentFuturePastType(blockWaiver) === CurrentFuturePastType.Current
  )

/**
 * has waiver?      | ghost       | vehicle and late indicator ghosts
 * ---------------- | ----------- | ---------------------------------
 * yes, current     | black       | black
 * yes, not current | highlighted | grey
 * none             | highlighted | none
 */
export const blockWaiverAlertStyle = (
  vehicleOrGhost: VehicleOrGhost
): AlertIconStyle | undefined => {
  if (hasCurrentBlockWaiver(vehicleOrGhost)) {
    return AlertIconStyle.Black
  }
  if (isGhost(vehicleOrGhost) && !isLateVehicleIndicator(vehicleOrGhost)) {
    return AlertIconStyle.Highlighted
  } else {
    return hasBlockWaiver(vehicleOrGhost) ? AlertIconStyle.Grey : undefined
  }
}
