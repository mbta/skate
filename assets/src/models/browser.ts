export const reload = (forceGet: boolean = false) => {
  window.location.reload(forceGet)
}
