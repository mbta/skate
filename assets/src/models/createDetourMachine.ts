import { setup, ActorLogicFrom, InputFrom } from "xstate"
import { Route, RoutePattern } from "../schedule"

export const createDetourMachine = setup({
  types: {} as {
    context: {
      route: Route
      routePattern: RoutePattern
    }

    input: {
      // Caller has target route pattern
      route: Route
      routePattern: RoutePattern
    }
    events: { type: "detour.edit.done" } | { type: "detour.edit.resume" }
  },
}).createMachine({
  id: "Detours Machine",
  context: ({ input }) => ({
    ...input,
  }),

  initial: "Detour Drawing",
  states: {
    "Detour Drawing": {
      initial: "Editing",
      states: {
        Editing: {
          on: {
            "detour.edit.done": {
              target: "Share Detour",
            },
          },
        },
        "Share Detour": {
          on: {
            "detour.edit.resume": {
              target: "Editing",
            },
          },
        },
      },
    },
  },
})

/**
 * This refers to the type of `input` provided in
 * {@linkcode createDetourMachine}'s {@linkcode setup} call
 */
export type CreateDetourMachineInput = InputFrom<
  ActorLogicFrom<typeof createDetourMachine>
>
