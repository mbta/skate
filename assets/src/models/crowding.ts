import { Vehicle } from "../realtime"

export type OccupancyStatus =
  | "NO_DATA"
  | "EMPTY"
  | "MANY_SEATS_AVAILABLE"
  | "FEW_SEATS_AVAILABLE"
  | "FULL"

export interface Crowding {
  load: number | null
  capacity: number | null
  occupancyStatus: OccupancyStatus
  occupancyPercentage: number | null
}

export const classModifierForStatus = (status: OccupancyStatus): string => {
  switch (status) {
    case "NO_DATA":
      return "no-data"
    case "EMPTY":
      return "empty"
    case "MANY_SEATS_AVAILABLE":
      return "not-crowded"
    case "FEW_SEATS_AVAILABLE":
      return "some-crowding"
    case "FULL":
      return "crowded"
  }
}

export const crowdingLabel = (vehicle: Vehicle): string => {
  if (vehicle.crowding && vehicle.crowding.load !== null) {
    return `${vehicle.crowding.load}/${vehicle.crowding.capacity}`
  } else {
    return "?/?"
  }
}

export const statusDescriptionForStatus = (status: OccupancyStatus): string => {
  switch (status) {
    case "NO_DATA":
      return "No data available"
    case "EMPTY":
      return "Empty"
    case "MANY_SEATS_AVAILABLE":
      return "Not crowded"
    case "FEW_SEATS_AVAILABLE":
      return "Some crowding"
    case "FULL":
      return "Crowded"
  }
}
