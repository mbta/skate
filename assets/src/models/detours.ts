export interface DraftDetour {
  route: string
  direction: string
  name: string
  intersection: string
  activeSince: number
}

export type DraftDetours = DraftDetour[]
