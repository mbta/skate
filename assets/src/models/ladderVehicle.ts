import { LadderDirection, TimepointStatusYFunc } from "../components/ladder"
import { Vehicle } from "../skate"

export interface LadderVehicle {
  vehicle: Vehicle
  x: number
  y: number
  vehicleDirection: VehicleDirection
  lane: number
}

export enum VehicleDirection {
  Down,
  Up,
}

interface WithVehicle {
  vehicle: Vehicle
}

interface OnLadder {
  y: number
  vehicleDirection: VehicleDirection
}

interface VehicleOnLadder extends WithVehicle, OnLadder {}

interface InLane extends OnLadder {
  lane: number
}

interface VehicleInLane extends WithVehicle, InLane {}

const widthOfVehicleGroup = 32
const heightOfVehicleGroup = 34

/**
 * The LadderVehicle data has some order dependency in how the properties are calculated.
 * E.g. In order to calculate x we need to know what lane the vehicle is in.
 *
 * Steps:
 *  1) Calculate VehicleOnLadder --> the properties that don't depend on anything else
 *  2) Calculate the lane property
 *  3) Calculate the x property
 */
export const ladderVehiclesFromVehicles = (
  vehicles: Vehicle[],
  ladderDirection: LadderDirection,
  timepointStatusYFunc: TimepointStatusYFunc
): {
  ladderVehicles: LadderVehicle[]
  widthOfLanes: number
} => {
  const vehiclesOnLadder: VehicleOnLadder[] = vehicles.map(vehicle =>
    vehicleOnLadder(vehicle, ladderDirection, timepointStatusYFunc)
  )

  const vehiclesInLane: VehicleInLane[] = putIntoLanes(vehiclesOnLadder)

  const maxOccupiedLane: number = numOccupiedLanes(vehiclesInLane)
  const ladderVehicleHorizontalOffset: number = horizontalOffsetForLanes(
    maxOccupiedLane
  )
  const xInLane = xInLaneWithOffset(ladderVehicleHorizontalOffset)
  const widthOfLanes =
    widthOfVehicleGroup + (maxOccupiedLane - 1) * ladderVehicleHorizontalOffset

  const ladderVehicles: LadderVehicle[] = vehiclesInLane.map(vehicleInLane =>
    addX(vehicleInLane, xInLane)
  )

  return {
    ladderVehicles,
    widthOfLanes,
  }
}

export const vehicleDirectionOnLadder = (
  vehicle: Vehicle,
  ladderDirection: LadderDirection
): VehicleDirection =>
  (vehicle.directionId === 1) ===
  (ladderDirection === LadderDirection.ZeroToOne)
    ? VehicleDirection.Down
    : VehicleDirection.Up

const vehicleOnLadder = (
  vehicle: Vehicle,
  ladderDirection: LadderDirection,
  timepointStatusYFunc: TimepointStatusYFunc
): VehicleOnLadder => {
  const vehicleDirection: VehicleDirection = vehicleDirectionOnLadder(
    vehicle,
    ladderDirection
  )

  const y = timepointStatusYFunc(vehicle.timepointStatus, vehicleDirection)

  return {
    // tslint:disable-next-line:object-literal-sort-keys
    vehicle,
    vehicleDirection,
    y,
  }
}

export const putIntoLanes = (
  vehiclesOnLadder: VehicleOnLadder[]
): VehicleInLane[] =>
  vehiclesOnLadder
    .slice()
    .sort(byDirectionAndY)
    .reduce(
      (
        previousLadderVehiclePositions: VehicleInLane[],
        ladderVehicle: VehicleOnLadder
      ): VehicleInLane[] => {
        const overlappers: InLane[] = overlappingPreviousVehicles(
          previousLadderVehiclePositions,
          ladderVehicle
        )
        const occupiedLanes: number[] = overlappers.map(
          overlapper => overlapper.lane
        )
        const lane = firstOpenLane(occupiedLanes)

        return [
          {
            ...ladderVehicle,
            // tslint:disable-next-line:object-literal-sort-keys
            lane,
          },
          ...previousLadderVehiclePositions,
        ]
      },
      []
    )

const numOccupiedLanes = (vehicles: InLane[]): number => {
  const allLanes: number[] = vehicles.map(({ lane }) => lane || 0)
  const furthestLane = allLanes.length ? Math.max(...allLanes) : 0
  return furthestLane + 1
}

const horizontalOffsetForLanes = (maxOccupiedLane: number): number => {
  const cappedLanes = Math.min(maxOccupiedLane, 5)
  return ((8 - cappedLanes) / 8) * widthOfVehicleGroup
}

const xInLaneWithOffset = (ladderVehicleHorizontalOffset: number) => (
  { vehicleDirection, lane }: InLane,
  baseX: number
): number => {
  const laneOffset = (lane || 0) * ladderVehicleHorizontalOffset
  return vehicleDirection === VehicleDirection.Up
    ? baseX + laneOffset
    : baseX - laneOffset
}

const addX = (
  vehicleInLane: VehicleInLane,
  xInLane: (vehicle: InLane, baseX: number) => number
): LadderVehicle => {
  const baseX =
    vehicleInLane.vehicleDirection === VehicleDirection.Up ? 63 : -63

  const x = xInLane(vehicleInLane, baseX)

  return {
    ...vehicleInLane,
    x,
  }
}

/**
 * Sort vehicles based on which direction they are traveling,
 * and given that direction, which vehicle is following the other.
 * E.g. traveling Down a smaller Y means you are behind.
 *
 * Sort order is only important for the internal logic of putIntoLanes.
 */
export const byDirectionAndY = (a: OnLadder, b: OnLadder): number => {
  if (a.vehicleDirection === VehicleDirection.Down) {
    if (b.vehicleDirection === VehicleDirection.Up) {
      return -1
    } else {
      if (a.y > b.y) {
        return -1
      } else if (b.y > a.y) {
        return 1
      } else {
        return 0
      }
    }
  } else {
    if (b.vehicleDirection === VehicleDirection.Down) {
      return 1
    } else {
      if (a.y < b.y) {
        return -1
      } else if (b.y < a.y) {
        return 1
      } else {
        return 0
      }
    }
  }
}

const overlappingPreviousVehicles = (
  previousVehicles: InLane[],
  vehicle: OnLadder
): InLane[] =>
  previousVehicles.filter(previousVehicle =>
    areOverlapping(previousVehicle, vehicle)
  )

export const areOverlapping = (a: OnLadder, b: OnLadder): boolean =>
  a.vehicleDirection === b.vehicleDirection &&
  Math.abs(a.y - b.y) < heightOfVehicleGroup

export const firstOpenLane = (occupiedLanes: number[]): number => {
  const sortedOccupiedLanes = occupiedLanes.slice().sort()

  for (let index = 0; index < sortedOccupiedLanes.length; index++) {
    if (sortedOccupiedLanes[index] !== index) {
      return index
    }
  }

  return sortedOccupiedLanes.length
}
