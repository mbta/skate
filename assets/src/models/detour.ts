import { LatLngLiteral } from "leaflet"
import { ShapePoint, Stop } from "../schedule"
import { CreateDetourMachineInput } from "./createDetourMachine"
import { any, Infer, number, string, type } from "superstruct"

export interface DetourWithState {
  author: string
  state: any
  updatedAt: number
}

export const DetourWithStateData = type({
  author: string(),
  state: any(),
  updated_at: number(),
})

export type DetourWithStateData = Infer<typeof DetourWithStateData>

export const detourStateFromData = (
  detourWithState: DetourWithStateData
): DetourWithState => ({
  author: detourWithState.author,
  state: detourWithState.state,
  updatedAt: detourWithState.updated_at,
})

export interface DetourShape {
  coordinates: ShapePoint[]
  directions?: DetourDirection[]
}

export type DetourDirection = {
  instruction: string
}

export interface MapLocation {
  center?: LatLngLiteral
  zoom?: number
}

export type OriginalRoute = CreateDetourMachineInput & MapLocation

export interface UnfinishedRouteSegments {
  beforeStartPoint: ShapePoint[]
  afterStartPoint: ShapePoint[]
}

export interface UnfinishedDetour {
  unfinishedRouteSegments: UnfinishedRouteSegments
}

export interface RouteSegments {
  beforeDetour: ShapePoint[]
  detour: ShapePoint[]
  afterDetour: ShapePoint[]
}

export interface FinishedDetour {
  missedStops: Stop[]
  connectionPoint: {
    start?: Stop
    end?: Stop
  }
  routeSegments: RouteSegments
  detourShape: DetourShape
}
