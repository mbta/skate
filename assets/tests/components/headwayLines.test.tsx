import React from "react"
import renderer from "react-test-renderer"
import HeadwayLines from "../../src/components/headwayLines"
import { LadderVehicle, VehicleDirection } from "../../src/models/ladderVehicle"

describe("HeadwayLines", () => {
  test("renders", () => {
    const vehicles: LadderVehicle[] = [
      {
        vehicleId: "1",
        label: "1",
        viaVariant: null,
        status: "on-time",
        headwaySpacing: "ok",
        isOffCourse: false,
        x: 1,
        y: 1,
        vehicleDirection: VehicleDirection.Up,
        lane: 0,
      },
      {
        vehicleId: "2",
        label: "2",
        viaVariant: null,
        status: "on-time",
        headwaySpacing: "bunched",
        isOffCourse: false,
        x: 1,
        y: 100,
        vehicleDirection: VehicleDirection.Up,
        lane: 0,
      },
      {
        vehicleId: "3",
        label: "3",
        viaVariant: null,
        status: "on-time",
        headwaySpacing: "gapped",
        isOffCourse: false,
        x: 1,
        y: 300,
        vehicleDirection: VehicleDirection.Down,
        lane: 0,
      },
    ]
    const tree = renderer
      .create(<HeadwayLines height={800} ladderVehicles={vehicles} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
