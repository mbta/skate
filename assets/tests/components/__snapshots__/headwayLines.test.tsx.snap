// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadwayLines renders 1`] = `
<g
  className="m-ladder__headway-lines"
>
  <line
    className="m-ladder__line
                  m-ladder__headway-line
                  m-ladder__headway-line--gapped"
    id="3"
    x1={-40}
    x2={-40}
    y1={800}
    y2={300}
  />
  <line
    className="m-ladder__line
                  m-ladder__headway-line
                  m-ladder__headway-line--bunched"
    id="2"
    x1={40}
    x2={40}
    y1={0}
    y2={100}
  />
  <line
    className="m-ladder__line
                  m-ladder__headway-line
                  m-ladder__headway-line--ok"
    id="1"
    x1={40}
    x2={40}
    y1={100}
    y2={1}
  />
</g>
`;
